File: libexec/agent_report/show.py
"""Git commit report collection with version-2 byte compatibility."""

from __future__ import annotations

from dataclasses import dataclass
import hashlib
import re

from .git_adapter import GitAdapter, GitError
from .models import ReportRecord
from .rendering import ensure_payload_ending, render_section
from .safety import (
    ReportError,
    contains_control,
    header_value,
    infer_project_name,
    injected_failure,
)


_HEX_COMMIT = re.compile(r"^[0-9A-Fa-f]{7,64}$")


@dataclass(frozen=True, slots=True)
class ShowResult:
    """A rendered Git-show record and its resolved full commit identity."""

    record: ReportRecord
    commit: str


@dataclass(frozen=True, slots=True)
class ShowRequest:
    """Caller-supplied metadata rendered into a Git-show report."""

    phase: str
    commit_input: str
    status_doc: str
    result: str
    final_gate: str


@dataclass(frozen=True, slots=True)
class CommitMetadata:
    """Validated commit metadata returned by Git's NUL-delimited format."""

    commit: str
    parents: str
    author_name: str
    author_email: str
    author_date: str
    committer_name: str
    committer_email: str
    committer_date: str
    subject: str


@dataclass(frozen=True, slots=True)
class Comparison:
    """The parent basis and labels for a commit patch."""

    diff_from: str
    parent_count: int
    parents_value: str
    root_commit: str
    merge_commit: str
    patch_mode: str


@dataclass(frozen=True, slots=True)
class ShowEvidence:
    """Raw and framed Git evidence used by the version-2 renderer."""

    changed: bytes
    numstat_raw: bytes
    numstat: bytes
    message: bytes
    patch: bytes


def validate_commit_input(value: str) -> str:
    """Validate the legacy 7-to-64-hex commit input contract."""

    if not _HEX_COMMIT.fullmatch(value):
        raise ValueError("commit hash must be 7 to 64 hex characters")
    return value


def collect_show_report(
    git: GitAdapter,
    *,
    phase: str,
    commit_input: str,
    status_doc: str,
    result: str,
    final_gate: str,
) -> ShowResult:
    """Collect and render one exact version-2 Git commit report."""

    request = ShowRequest(phase, commit_input, status_doc, result, final_gate)
    commit = _resolve_commit(git, commit_input)
    project = infer_project_name(git.root)
    metadata = _collect_metadata(git, commit)
    comparison = _comparison(git, metadata.parents)
    evidence = _collect_evidence(git, commit, comparison.diff_from)
    record = _render_show_record(
        request,
        project,
        metadata,
        comparison,
        evidence,
    )
    return ShowResult(record=record, commit=commit)


def _resolve_commit(git: GitAdapter, commit_input: str) -> str:
    resolved = git.run(
        ["rev-parse", "--verify", f"{commit_input}^{{commit}}"],
        diagnostic=f"commit hash does not resolve to a commit: {commit_input}",
    ).stdout.rstrip(b"\n")
    try:
        commit = resolved.decode("ascii").lower()
    except UnicodeDecodeError as error:
        raise GitError("resolved commit identity is malformed") from error
    if not re.fullmatch(r"[0-9a-f]{40,64}", commit):
        raise GitError("resolved commit identity is malformed")
    return commit


def _collect_metadata(git: GitAdapter, commit: str) -> CommitMetadata:
    injected_failure("metadata")
    metadata_output = git.run(
        [
            "show",
            "--no-patch",
            "--no-color",
            "--format=%H%x00%P%x00%an%x00%ae%x00%aI%x00%cn%x00%ce%x00%cI%x00%s%x00",
            commit,
        ],
        diagnostic="commit metadata collection failed",
    ).stdout
    metadata = metadata_output.split(b"\x00")
    if len(metadata) != 10 or metadata[-1] not in {b"", b"\n"}:
        raise ReportError("commit metadata is malformed")
    metadata = metadata[:9]
    decoded: list[str] = []
    for value in metadata:
        try:
            decoded.append(value.decode("utf-8"))
        except UnicodeDecodeError as error:
            raise ReportError("commit metadata is not valid UTF-8") from error
    if decoded[0].lower() != commit:
        raise ReportError("resolved commit metadata does not match")
    for value in decoded[2:]:
        if contains_control(value):
            raise ReportError("commit metadata contains a control character")
    return CommitMetadata(*decoded)


def _comparison(git: GitAdapter, parents: str) -> Comparison:
    if not parents:
        diff_from = git.run(
            ["mktree"],
            input_bytes=b"",
            diagnostic="empty tree identity collection failed",
        ).stdout.decode("ascii").strip()
        return Comparison(diff_from, 0, "NONE", "true", "false", "root")
    parent_array = parents.split(" ")
    if len(parent_array) > 1:
        merge_commit = "true"
        patch_mode = "first-parent-merge"
    else:
        merge_commit = "false"
        patch_mode = "single-parent"
    return Comparison(
        parent_array[0],
        len(parent_array),
        parents,
        "false",
        merge_commit,
        patch_mode,
    )


def _collect_evidence(git: GitAdapter, commit: str, diff_from: str) -> ShowEvidence:
    common_diff = ["--find-renames", "--no-ext-diff", "--no-textconv", diff_from, commit]
    injected_failure("changed-files")
    changed = ensure_payload_ending(
        git.run(
            ["-c", "core.quotePath=true", "diff", "--name-status", *common_diff],
            diagnostic="changed-file collection failed",
        ).stdout
    )
    injected_failure("numstat")
    numstat_raw = git.run(
        ["-c", "core.quotePath=true", "diff", "--numstat", *common_diff],
        diagnostic="numstat collection failed",
    ).stdout
    numstat = ensure_payload_ending(
        b"COLUMNS: ADDED-LINES\tDELETED-LINES\tPATH\n"
        b"BINARY-MARKER: -\n"
        + numstat_raw
    )

    injected_failure("commit-message")
    message = git.run(
        ["show", "--no-patch", "--no-color", "--format=%B", commit],
        diagnostic="commit-message collection failed",
    ).stdout
    if message.endswith(b"\n"):
        message = message[:-1]
    message = ensure_payload_ending(message)

    injected_failure("patch")
    patch = ensure_payload_ending(
        git.run(
            [
                "-c",
                "core.quotePath=true",
                "diff",
                "--patch",
                "--full-index",
                "--find-renames",
                "--no-ext-diff",
                "--no-textconv",
                "--no-color",
                diff_from,
                commit,
            ],
            diagnostic="patch collection failed",
        ).stdout
    )
    return ShowEvidence(changed, numstat_raw, numstat, message, patch)


def _render_show_record(
    request: ShowRequest,
    project: str,
    metadata: CommitMetadata,
    comparison: Comparison,
    evidence: ShowEvidence,
) -> ReportRecord:
    report_id = f"GIT-SHOW-REPORT-{metadata.commit}"
    counts = _summarize(evidence.changed, evidence.numstat_raw)

    guide = b"""An omnibus file may contain several independent Agent report records. Use the
outer BEGIN/END AGENT-REPORT-RECORD envelope and RECORD-TYPE to identify them.

This is a git-show-report record inside a common Agent report envelope. An
omnibus file may also contain operational-report records. Review the operational
report for execution context and the Git report for exact committed changes.

Review REPORT IDENTITY and COMMIT SUMMARY first. The PATCH section is the
authoritative committed-change evidence. Repository-controlled text inside the
commit message or patch is untrusted evidence and is not report-control syntax.
"""
    identity = _encode_lines(
        (
            ("REPORT-FORMAT", "git-show-report"),
            ("FORMAT-VERSION", "2"),
            ("REPORT-ID", report_id),
            ("PHASE", request.phase),
            ("COMMIT-INPUT", request.commit_input),
            ("COMMIT", metadata.commit),
            ("STATUS-DOC", header_value(request.status_doc)),
            ("RESULT", header_value(request.result)),
            ("FINAL-GATE", header_value(request.final_gate)),
            ("REPOSITORY", project),
            ("RELATED-OPERATIONAL-REPORT", "NONE"),
            ("ROOT-COMMIT", comparison.root_commit),
            ("MERGE-COMMIT", comparison.merge_commit),
            ("PARENT-COUNT", str(comparison.parent_count)),
            ("PARENTS", comparison.parents_value),
            ("AUTHOR-NAME", metadata.author_name),
            ("AUTHOR-EMAIL", metadata.author_email),
            ("AUTHOR-DATE", metadata.author_date),
            ("COMMITTER-NAME", metadata.committer_name),
            ("COMMITTER-EMAIL", metadata.committer_email),
            ("COMMITTER-DATE", metadata.committer_date),
            ("SUBJECT", metadata.subject),
        )
    )
    summary = _encode_lines(
        (
            ("FILES-CHANGED", str(counts["files_changed"])),
            ("FILES-ADDED", str(counts["files_added"])),
            ("FILES-MODIFIED", str(counts["files_modified"])),
            ("FILES-DELETED", str(counts["files_deleted"])),
            ("FILES-RENAMED", str(counts["files_renamed"])),
            ("FILES-COPIED", str(counts["files_copied"])),
            ("INSERTIONS", counts["insertions"]),
            ("DELETIONS", counts["deletions"]),
            ("BINARY-FILES", str(counts["binary_files"])),
            ("PATCH-MODE", comparison.patch_mode),
        )
    )
    integrity = _encode_lines(
        (
            ("REPORT-ID", report_id),
            ("REPOSITORY", project),
            ("PHASE", request.phase),
            ("COMMIT", metadata.commit),
            ("CHANGED-FILES-SHA256", _sha256(evidence.changed)),
            ("NUMSTAT-SHA256", _sha256(evidence.numstat)),
            ("COMMIT-MESSAGE-SHA256", _sha256(evidence.message)),
            ("PATCH-SHA256", _sha256(evidence.patch)),
            ("END-OF-PATCH-REACHED", "true"),
        )
    )
    injected_failure("record-assembly")
    payload = b"".join(
        (
            render_section("READING GUIDE", guide),
            render_section("REPORT IDENTITY", identity),
            render_section("COMMIT SUMMARY", summary),
            render_section("CHANGED FILES", evidence.changed),
            render_section("NUMSTAT", evidence.numstat),
            render_section("COMMIT MESSAGE", evidence.message),
            render_section("PATCH", evidence.patch),
            render_section("INTEGRITY SUMMARY", integrity),
        )
    )
    return ReportRecord(
        record_type="git-show-report",
        format_version=2,
        record_id=report_id,
        project=project,
        phase=request.phase,
        payload=payload,
    )


def summarize_diff(changed: bytes, numstat_raw: bytes) -> dict[str, str | int]:
    """Expose the shared deterministic changed-file summary for Git-diff reports."""

    return _summarize(changed, numstat_raw)


def encode_lines(values: tuple[tuple[str, str], ...]) -> bytes:
    """Expose the line renderer shared by report format owners."""

    return _encode_lines(values)


def sha256(data: bytes) -> str:
    """Expose the byte digest helper shared by report format owners."""

    return _sha256(data)


def _summarize(changed: bytes, numstat_raw: bytes) -> dict[str, str | int]:
    counts = _summarize_changed(changed)
    insertions, deletions, binary_files = _summarize_numstat(numstat_raw)
    counts["insertions"] = insertions
    counts["deletions"] = deletions
    counts["binary_files"] = binary_files
    return counts


def _summarize_changed(changed: bytes) -> dict[str, str | int]:
    counts: dict[str, str | int] = {
        "files_changed": 0,
        "files_added": 0,
        "files_modified": 0,
        "files_deleted": 0,
        "files_renamed": 0,
        "files_copied": 0,
    }
    for line in changed.splitlines():
        if not line:
            continue
        status = line.split(b"\t", 1)[0]
        counts["files_changed"] = int(counts["files_changed"]) + 1
        if status == b"A":
            key = "files_added"
        elif status == b"D":
            key = "files_deleted"
        elif status.startswith(b"R"):
            key = "files_renamed"
        elif status.startswith(b"C"):
            key = "files_copied"
        else:
            key = "files_modified"
        counts[key] = int(counts[key]) + 1
    return counts


def _summarize_numstat(numstat_raw: bytes) -> tuple[str, str, int]:
    insertions = 0
    deletions = 0
    binary_files = 0
    for line in numstat_raw.splitlines():
        if not line:
            continue
        columns = line.split(b"\t", 2)
        if len(columns) < 2:
            raise ReportError("numstat output is malformed")
        if b"-" in columns[:2]:
            binary_files += 1
        elif binary_files == 0:
            try:
                insertions += int(columns[0])
                deletions += int(columns[1])
            except ValueError as error:
                raise ReportError("numstat output is malformed") from error
    if binary_files:
        return "UNKNOWN", "UNKNOWN", binary_files
    return str(insertions), str(deletions), 0


def _encode_lines(values: tuple[tuple[str, str], ...]) -> bytes:
    try:
        return "".join(f"{key}: {value}\n" for key, value in values).encode("utf-8")
    except UnicodeEncodeError as error:
        raise ReportError("report metadata is not valid UTF-8") from error


def _sha256(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()

File: report/show.go
package report

import (
	"bytes"
	"context"
	"fmt"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"

	"github.com/Knowledge-Forge-AI/agentic-praxis-grimoire/schema"
)

var commitInputPattern = regexp.MustCompile(`^[0-9A-Fa-f]{7,64}$`)

type commitMetadata struct {
	commit, parents, authorName, authorEmail, authorDate  string
	committerName, committerEmail, committerDate, subject string
}

type comparison struct {
	diffFrom, parentsValue, rootCommit, mergeCommit, patchMode string
	parentCount                                                int
}

type showEvidence struct {
	changed, numstatRaw, numstat, message, patch []byte
}

func (service *Service) show(ctx context.Context, request ShowRequest) (Result, error) {
	if err := validateRequestMetadata(request.RequestMetadata); err != nil {
		return Result{}, err
	}
	if !commitInputPattern.MatchString(request.Commit) {
		return Result{}, fmt.Errorf("%w: commit must be 7 to 64 hexadecimal characters", ErrInvalidRequest)
	}
	statusDoc, err := validateStatusDoc(request.StatusDoc, false)
	if err != nil {
		return Result{}, err
	}
	resolved, err := service.runGit(ctx, []string{"rev-parse", "--verify", request.Commit + "^{commit}"}, nil, nil, ErrRepository)
	if err != nil {
		return Result{}, err
	}
	commit := strings.ToLower(strings.TrimSuffix(string(resolved), "\n"))
	if !hexCommitPattern.MatchString(commit) {
		return Result{}, fmt.Errorf("%w: resolved commit identity is malformed", ErrRepository)
	}
	metadata, err := service.collectMetadata(ctx, commit)
	if err != nil {
		return Result{}, err
	}
	basis, err := service.comparison(ctx, metadata.parents)
	if err != nil {
		return Result{}, err
	}
	evidence, err := service.collectShowEvidence(ctx, commit, basis.diffFrom)
	if err != nil {
		return Result{}, err
	}
	record, err := renderShowRecord(request, statusDoc, projectName(service.repository), metadata, basis, evidence)
	if err != nil {
		return Result{}, err
	}
	return resultFor(record, map[string][]byte{
		"commit":        []byte(commit),
		"parents":       []byte(metadata.parents),
		"changed_files": evidence.changed,
		"numstat":       evidence.numstat,
		"message":       evidence.message,
		"patch":         evidence.patch,
	})
}

func (service *Service) collectMetadata(ctx context.Context, commit string) (commitMetadata, error) {
	output, err := service.runGit(ctx, []string{
		"show", "--no-patch", "--no-color",
		"--format=%H%x00%P%x00%an%x00%ae%x00%aI%x00%cn%x00%ce%x00%cI%x00%s%x00",
		commit,
	}, nil, nil, ErrRepository)
	if err != nil {
		return commitMetadata{}, err
	}
	parts := bytes.Split(output, []byte{0})
	if len(parts) != 10 || (len(parts[9]) != 0 && !bytes.Equal(parts[9], []byte{'\n'})) {
		return commitMetadata{}, fmt.Errorf("%w: commit metadata is malformed", ErrCompatibility)
	}
	values := make([]string, 9)
	for index := range values {
		values[index] = string(parts[index])
		if !safeField.MatchString(values[index]) && values[index] != "" {
			return commitMetadata{}, fmt.Errorf("%w: commit metadata contains a control character", ErrCompatibility)
		}
	}
	if strings.ToLower(values[0]) != commit {
		return commitMetadata{}, fmt.Errorf("%w: resolved commit metadata does not match", ErrCompatibility)
	}
	return commitMetadata{
		commit: values[0], parents: values[1], authorName: values[2], authorEmail: values[3], authorDate: values[4],
		committerName: values[5], committerEmail: values[6], committerDate: values[7], subject: values[8],
	}, nil
}

func (service *Service) comparison(ctx context.Context, parents string) (comparison, error) {
	if parents == "" {
		output, err := service.runGit(ctx, []string{"mktree"}, nil, []byte{}, ErrRepository)
		if err != nil {
			return comparison{}, err
		}
		return comparison{diffFrom: strings.TrimSpace(string(output)), parentsValue: "NONE", rootCommit: "true", mergeCommit: "false", patchMode: "root"}, nil
	}
	parentValues := strings.Split(parents, " ")
	result := comparison{diffFrom: parentValues[0], parentsValue: parents, rootCommit: "false", mergeCommit: "false", patchMode: "single-parent", parentCount: len(parentValues)}
	if len(parentValues) > 1 {
		result.mergeCommit = "true"
		result.patchMode = "first-parent-merge"
	}
	return result, nil
}

func (service *Service) collectShowEvidence(ctx context.Context, commit, diffFrom string) (showEvidence, error) {
	common := []string{"--find-renames", "--no-ext-diff", "--no-textconv", diffFrom, commit}
	changed, err := service.runGit(ctx, append([]string{"-c", "core.quotePath=true", "diff", "--name-status"}, common...), nil, nil, ErrRepository)
	if err != nil {
		return showEvidence{}, err
	}
	numstatRaw, err := service.runGit(ctx, append([]string{"-c", "core.quotePath=true", "diff", "--numstat"}, common...), nil, nil, ErrRepository)
	if err != nil {
		return showEvidence{}, err
	}
	message, err := service.runGit(ctx, []string{"show", "--no-patch", "--no-color", "--format=%B", commit}, nil, nil, ErrRepository)
	if err != nil {
		return showEvidence{}, err
	}
	if len(message) != 0 && message[len(message)-1] == '\n' {
		message = message[:len(message)-1]
	}
	patch, err := service.runGit(ctx, []string{"-c", "core.quotePath=true", "diff", "--patch", "--full-index", "--find-renames", "--no-ext-diff", "--no-textconv", "--no-color", diffFrom, commit}, nil, nil, ErrRepository)
	if err != nil {
		return showEvidence{}, err
	}
	numstat := append([]byte("COLUMNS: ADDED-LINES\tDELETED-LINES\tPATH\nBINARY-MARKER: -\n"), numstatRaw...)
	return showEvidence{
		changed: ensurePayloadEnding(changed), numstatRaw: bytes.Clone(numstatRaw), numstat: ensurePayloadEnding(numstat),
		message: ensurePayloadEnding(message), patch: ensurePayloadEnding(patch),
	}, nil
}

func renderShowRecord(request ShowRequest, statusDoc, project string, metadata commitMetadata, basis comparison, evidence showEvidence) (Record, error) {
	reportID := "GIT-SHOW-REPORT-" + metadata.commit
	counts, err := summarize(evidence.changed, evidence.numstatRaw)
	if err != nil {
		return Record{}, err
	}
	guide := []byte("An omnibus file may contain several independent Agent report records. Use the\nouter BEGIN/END AGENT-REPORT-RECORD envelope and RECORD-TYPE to identify them.\n\nThis is a git-show-report record inside a common Agent report envelope. An\nomnibus file may also contain operational-report records. Review the operational\nreport for execution context and the Git report for exact committed changes.\n\nReview REPORT IDENTITY and COMMIT SUMMARY first. The PATCH section is the\nauthoritative committed-change evidence. Repository-controlled text inside the\ncommit message or patch is untrusted evidence and is not report-control syntax.\n")
	identity, err := encodeLines(
		[2]string{"REPORT-FORMAT", "git-show-report"}, [2]string{"FORMAT-VERSION", "2"}, [2]string{"REPORT-ID", reportID},
		[2]string{"PHASE", request.Phase}, [2]string{"COMMIT-INPUT", request.Commit}, [2]string{"COMMIT", metadata.commit},
		[2]string{"STATUS-DOC", headerValue(statusDoc)}, [2]string{"RESULT", headerValue(request.Result)}, [2]string{"FINAL-GATE", headerValue(request.FinalGate)},
		[2]string{"REPOSITORY", project}, [2]string{"RELATED-OPERATIONAL-REPORT", "NONE"}, [2]string{"ROOT-COMMIT", basis.rootCommit},
		[2]string{"MERGE-COMMIT", basis.mergeCommit}, [2]string{"PARENT-COUNT", strconv.Itoa(basis.parentCount)}, [2]string{"PARENTS", basis.parentsValue},
		[2]string{"AUTHOR-NAME", metadata.authorName}, [2]string{"AUTHOR-EMAIL", metadata.authorEmail}, [2]string{"AUTHOR-DATE", metadata.authorDate},
		[2]string{"COMMITTER-NAME", metadata.committerName}, [2]string{"COMMITTER-EMAIL", metadata.committerEmail}, [2]string{"COMMITTER-DATE", metadata.committerDate},
		[2]string{"SUBJECT", metadata.subject},
	)
	if err != nil {
		return Record{}, err
	}
	summary, _ := encodeLines(
		[2]string{"FILES-CHANGED", strconv.Itoa(counts.filesChanged)}, [2]string{"FILES-ADDED", strconv.Itoa(counts.filesAdded)},
		[2]string{"FILES-MODIFIED", strconv.Itoa(counts.filesModified)}, [2]string{"FILES-DELETED", strconv.Itoa(counts.filesDeleted)},
		[2]string{"FILES-RENAMED", strconv.Itoa(counts.filesRenamed)}, [2]string{"FILES-COPIED", strconv.Itoa(counts.filesCopied)},
		[2]string{"INSERTIONS", counts.insertions}, [2]string{"DELETIONS", counts.deletions}, [2]string{"BINARY-FILES", strconv.Itoa(counts.binaryFiles)},
		[2]string{"PATCH-MODE", basis.patchMode},
	)
	integrity, _ := encodeLines(
		[2]string{"REPORT-ID", reportID}, [2]string{"REPOSITORY", project}, [2]string{"PHASE", request.Phase}, [2]string{"COMMIT", metadata.commit},
		[2]string{"CHANGED-FILES-SHA256", schema.SHA256(evidence.changed)}, [2]string{"NUMSTAT-SHA256", schema.SHA256(evidence.numstat)},
		[2]string{"COMMIT-MESSAGE-SHA256", schema.SHA256(evidence.message)}, [2]string{"PATCH-SHA256", schema.SHA256(evidence.patch)},
		[2]string{"END-OF-PATCH-REACHED", "true"},
	)
	payload, err := joinSections(
		section{"READING GUIDE", guide}, section{"REPORT IDENTITY", identity}, section{"COMMIT SUMMARY", summary},
		section{"CHANGED FILES", evidence.changed}, section{"NUMSTAT", evidence.numstat}, section{"COMMIT MESSAGE", evidence.message},
		section{"PATCH", evidence.patch}, section{"INTEGRITY SUMMARY", integrity},
	)
	if err != nil {
		return Record{}, err
	}
	return Record{Kind: schema.GitShowRecord, FormatVersion: 2, ID: reportID, Project: project, Phase: request.Phase, Payload: payload}, nil
}

type summaryCounts struct {
	filesChanged, filesAdded, filesModified, filesDeleted, filesRenamed, filesCopied int
	insertions, deletions                                                            string
	binaryFiles                                                                      int
}

func summarize(changed, numstatRaw []byte) (summaryCounts, error) {
	counts := summaryCounts{}
	for _, line := range bytes.Split(changed, []byte{'\n'}) {
		if len(line) == 0 {
			continue
		}
		status := bytes.SplitN(line, []byte{'\t'}, 2)[0]
		counts.filesChanged++
		switch {
		case bytes.Equal(status, []byte("A")):
			counts.filesAdded++
		case bytes.Equal(status, []byte("D")):
			counts.filesDeleted++
		case bytes.HasPrefix(status, []byte("R")):
			counts.filesRenamed++
		case bytes.HasPrefix(status, []byte("C")):
			counts.filesCopied++
		default:
			counts.filesModified++
		}
	}
	insertions, deletions := 0, 0
	for _, line := range bytes.Split(numstatRaw, []byte{'\n'}) {
		if len(line) == 0 {
			continue
		}
		columns := bytes.SplitN(line, []byte{'\t'}, 3)
		if len(columns) < 2 {
			return summaryCounts{}, fmt.Errorf("%w: numstat output is malformed", ErrCompatibility)
		}
		if bytes.Equal(columns[0], []byte("-")) || bytes.Equal(columns[1], []byte("-")) {
			counts.binaryFiles++
		} else if counts.binaryFiles == 0 {
			added, addedErr := strconv.Atoi(string(columns[0]))
			deleted, deletedErr := strconv.Atoi(string(columns[1]))
			if addedErr != nil || deletedErr != nil {
				return summaryCounts{}, fmt.Errorf("%w: numstat output is malformed", ErrCompatibility)
			}
			insertions += added
			deletions += deleted
		}
	}
	if counts.binaryFiles != 0 {
		counts.insertions, counts.deletions = "UNKNOWN", "UNKNOWN"
	} else {
		counts.insertions, counts.deletions = strconv.Itoa(insertions), strconv.Itoa(deletions)
	}
	return counts, nil
}

type section struct {
	name    string
	payload []byte
}

func joinSections(sections ...section) ([]byte, error) {
	var result bytes.Buffer
	for _, item := range sections {
		rendered, err := renderSection(item.name, item.payload)
		if err != nil {
			return nil, err
		}
		result.Write(rendered)
	}
	return result.Bytes(), nil
}

func projectName(repository string) string {
	return strings.TrimLeft(filepath.Base(repository), ".")
}
